### PR TITLE
Add device based aspect ratio selection

### PR DESCRIPTION
**Pull request by belanagy**

### CODE REVIEW

File: sccameraview/src/main/java/io/supercharge/sccameraview/AspectRatio.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    @RequiresApi(Build.VERSION_CODES.LOLLIPOP)
+    public Size getSize() {
+        return new Size(width, height);
+    }
+
+    public Camera.Size getSize(Camera camera) {
+        return camera.new Size(width, height);
+    }
+
+    public String getRatioSize() {
```
Comment (richardradics) [inline sccameraview/src/main/java/io/supercharge/sccameraview/AspectRatio.java]: ?

File: sccameraview/build.gradle [hunk 2]
```diff
@@ -12,6 +12,22 @@ android {
...
+
+allprojects {
+    repositories {
+        jcenter()
+    }
 }
+
+dependencies {
+    compile "com.android.support:support-core-utils:25.3.1"
+    compile 'com.android.support.constraint:constraint-layout:1.0.2'
```
Comment (richardradics) [inline sccameraview/build.gradle]: do we need this dependency in the lib?

File: sccameraview/src/main/java/io/supercharge/sccameraview/AspectRatio.java [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+
+public class AspectRatio implements Serializable {
+
+    private double ratio;
+    private int width;
+    private int height;
+    private int x;
+    private int y;
+    private int gcd;
```
Comment (richardradics) [inline sccameraview/src/main/java/io/supercharge/sccameraview/AspectRatio.java]: gcd? i would give a more understandable fieldname

File: sccameraview/src/main/AndroidManifest.xml [hunk 4]
```diff
@@ -1,5 +1,5 @@
 <manifest xmlns:android="http://schemas.android.com/apk/res/android"
-          package="com.example.sccameraview">
+          package="io.supercharge.sccameraview">
 
     <application
          android:label="@string/app_name"
```
Comment (richardradics) [inline sccameraview/src/main/AndroidManifest.xml]: remove label, support rtsl

File: sccameraview/src/main/java/io/supercharge/sccameraview/AspectRatio.java [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+
+    private void calculateRatio(int width, int height) {
+        this.gcd = gcd(width, height);
+        if (gcd > 0) {
+            this.x = width / gcd;
+            this.y = height / gcd;
+        }
+    }
+
+    private static int gcd(int a, int b) {
```
Comment (richardradics) [inline sccameraview/src/main/java/io/supercharge/sccameraview/AspectRatio.java]: more understandable name pls

File: sccameraview/src/main/java/io/supercharge/sccameraview/Camera1View.java [hunk 6]
```diff
@@ -64,13 +59,14 @@ void openCamera() {
...
 
+        parameters.setRotation(ORIENTATION_270);
         camera.setDisplayOrientation(ORIENTATION_90);
         camera.setParameters(parameters);
         try {
             camera.setPreviewTexture(getSurfaceTexture());
             camera.startPreview();
         } catch (IOException e) {
-            Log.e(LOG_TAG, e.getMessage());
+            e.printStackTrace();
```
Comment (richardradics) [inline sccameraview/src/main/java/io/supercharge/sccameraview/Camera1View.java]: remove this, use `Log.e(LOG_TAG, e.getMessage());`

File: sccameraview/src/main/java/io/supercharge/sccameraview/Camera1View.java [hunk 7]
```diff
@@ -178,15 +180,18 @@ public void stopRecordingVideo() {
         } catch (RuntimeException e) {
             // RuntimeException is thrown when stop() is called immediately after start().
             // In this case the output file is not properly constructed ans should be deleted.
-            Log.e(LOG_TAG, e.getMessage());
+            e.printStackTrace();
```
Comment (richardradics) [inline sccameraview/src/main/java/io/supercharge/sccameraview/Camera1View.java]: remove this, use: `Log.e(LOG_TAG, e.getMessage());`

File: sccameraview/src/main/java/io/supercharge/sccameraview/Camera2View.java [hunk 8]
```diff
@@ -114,15 +104,14 @@ private void stopBackgroundThread() {
             backgroundThread = null;
             backgroundHandler = null;
         } catch (InterruptedException e) {
-            Log.e(LOG_TAG, e.getMessage());
+            e.printStackTrace();
```
Comment (richardradics) [inline sccameraview/src/main/java/io/supercharge/sccameraview/Camera2View.java]: remove this, use `Log.e(LOG_TAG, e.getMessage());`

File: sccameraview/src/main/java/io/supercharge/sccameraview/Camera2View.java [hunk 9]
```diff
@@ -144,33 +131,61 @@ public void openCamera() {
...
+            previewSize = ratioSizeList.get(selectedRatioIdx).getSize();
+            videoSize = ratioSizeList.get(selectedRatioIdx).getSize();
+            imageSize = ratioSizeList.get(selectedRatioIdx).getSize();
 
             sensorOrientation = characteristics.get(CameraCharacteristics.SENSOR_ORIENTATION);
             mediaRecorder = new MediaRecorder();
             cameraManager.openCamera(cameraIdString, stateCallback, null);
         } catch (CameraAccessException e) {
-            Log.e(LOG_TAG, e.getMessage());
+            e.printStackTrace();
```
Comment (richardradics) [inline sccameraview/src/main/java/io/supercharge/sccameraview/Camera2View.java]: remove this, use `Log.e(LOG_TAG, e.getMessage());`

File: sccameraview/src/main/java/io/supercharge/sccameraview/Camera2View.java [hunk 10]
```diff
@@ -144,33 +131,61 @@ public void openCamera() {
...
         } catch (NullPointerException e) {
             // Currently an NPE is thrown when the Camera2API is used but not supported on the
             // device this code runs.
-            Log.e(LOG_TAG, e.getMessage());
+            e.printStackTrace();
```
Comment (richardradics) [inline sccameraview/src/main/java/io/supercharge/sccameraview/Camera2View.java]: remove this, use `Log.e(LOG_TAG, e.getMessage());`

File: sccameraview/src/main/java/io/supercharge/sccameraview/Camera2View.java [hunk 11]
```diff
@@ -234,7 +249,7 @@ public void onConfigureFailed(@NonNull CameraCaptureSession session) {
                         }
                     }, backgroundHandler);
         } catch (CameraAccessException e) {
-            Log.e(LOG_TAG, e.getMessage());
+            e.printStackTrace();
```
Comment (richardradics) [inline sccameraview/src/main/java/io/supercharge/sccameraview/Camera2View.java]: remove this, use `Log.e(LOG_TAG, e.getMessage());`

File: sccameraview/src/main/java/io/supercharge/sccameraview/Camera2View.java [hunk 12]
```diff
@@ -248,7 +263,7 @@ private void updatePreview() {
             thread.start();
             previewSession.setRepeatingRequest(previewBuilder.build(), null, backgroundHandler);
         } catch (CameraAccessException e) {
-            Log.e(LOG_TAG, e.getMessage());
+            e.printStackTrace();
```
Comment (richardradics) [inline sccameraview/src/main/java/io/supercharge/sccameraview/Camera2View.java]: remove this, use `Log.e(LOG_TAG, e.getMessage());`

File: sccameraview/src/main/java/io/supercharge/sccameraview/Camera1View.java (hunk 6 above)
Comment (belanagy) [reply sccameraview/src/main/java/io/supercharge/sccameraview/Camera1View.java]: removed logs and unnecessary dependencies

File: sccameraview/src/main/java/io/supercharge/sccameraview/Camera2View.java [hunk 13]
```diff
@@ -311,7 +326,7 @@ public void onConfigureFailed(@NonNull CameraCaptureSession cameraCaptureSession
             }, backgroundHandler);
             recordingVideo = true;
         } catch (CameraAccessException | IOException e) {
-            Log.e(LOG_TAG, e.getMessage());
+            e.printStackTrace();
```
Comment (richardradics) [inline sccameraview/src/main/java/io/supercharge/sccameraview/Camera2View.java]: remove this, use `Log.e(LOG_TAG, e.getMessage());`

File: sccameraview/src/main/java/io/supercharge/sccameraview/SaveImageTask.java [hunk 14]
```diff
@@ -40,9 +32,9 @@ protected Object doInBackground(Object[] objects) {
                 fos.write(imageData);
                 fos.close();
             } catch (FileNotFoundException e) {
-                Log.d(LOG_TAG, "File not found: " + e.getMessage());
+                e.printStackTrace();
```
Comment (richardradics) [inline sccameraview/src/main/java/io/supercharge/sccameraview/SaveImageTask.java]: remove this, use `Log.e(LOG_TAG, e.getMessage());`

File: sccameraview/src/main/java/io/supercharge/sccameraview/SaveImageTask.java [hunk 15]
```diff
@@ -40,9 +32,9 @@ protected Object doInBackground(Object[] objects) {
...
             } catch (IOException e) {
-                Log.d(LOG_TAG, "Error accessing file: " + e.getMessage());
+                e.printStackTrace();
```
Comment (richardradics) [inline sccameraview/src/main/java/io/supercharge/sccameraview/SaveImageTask.java]: remove this, use `Log.e(LOG_TAG, e.getMessage());`

File: app/build.gradle [hunk 16]
```diff
@@ -13,9 +13,24 @@ android {
     }
 }
 
+buildscript {
```
Comment (WonderCsabo) [inline app/build.gradle]: These are already declared in the top-level project.

File: sccameraview/build.gradle [hunk 17]
```diff
@@ -12,6 +12,21 @@ android {
     }
 }
 
-dependencies {
-    compile "com.android.support:support-annotations:25.3.1"
+buildscript {
```
Comment (WonderCsabo) [inline sccameraview/build.gradle]: Again, these are already in the root project.

**Review by WonderCsabo — CHANGES_REQUESTED** [review]

**Review by WonderCsabo — APPROVED** [review]

**Pull request merged**